src/runtime/security/kinit_context.cpp

```diff
@@ -0,0 +1,306 @@
+    }
+
+    _user_name.assign((const char *)buf);
+    return error_s::ok();
+}
+
+error_s kinit_context_impl::get_credentials()
+{
+    krb5_creds creds;
+    error_s err = error_s::ok();
+
+    // get initial credentials using a key table
+    // Notice: the contents of a krb5_creds structure need to be freed by ourselves
+    err = wrap_krb5_err(krb5_get_init_creds_keytab(_krb5_context,
+                                                   &creds,
+                                                   _principal,
+                                                   _keytab,
+                                                   0 /*valid from now*/,
+                                                   nullptr /*empty TKT service name*/,
+                                                   _opt),
+                        "get_init_cred");
+    if (!err.is_ok()) {
+        dwarn_f("get credentials of {} from KDC failed, reason({})",
+                FLAGS_krb5_principal,
+                err.description());
+        return err;
+    }
+    auto cleanup = dsn::defer([&]() { krb5_free_cred_contents(_krb5_context, &creds); });
+
+    // store credentials into _ccache.
+    err = wrap_krb5_err(krb5_cc_store_cred(_krb5_context, _ccache, &creds), "store_cred");
+    if (!err.is_ok()) {
+        dwarn_f("store credentials of {} to cache failed, err({})",
+                FLAGS_krb5_principal,
+                err.description());
+        return err;
+    }
+
+    _cred_expire_timestamp = creds.times.endtime;
+    ddebug_f("get credentials of {} from KDC ok, expires at {}",
+             FLAGS_krb5_principal,
+             utils::time_s_to_date_time(_cred_expire_timestamp));
+    return err;
+}
+
+void kinit_context_impl::schedule_renew_credentials()
+{
+    // reserve 300 seconds for renew
+    int64_t renew_gap = _cred_expire_timestamp - utils::get_current_physical_time_s() - 300;
+    if (renew_gap < 300)
+        renew_gap = 300;
+    ddebug_f("schedule to renew credentials in {} seconds later", renew_gap);
+
+    // why don't we use timers in rDSN framework?
+    //  1. currently the rdsn framework may not started yet.
+    //  2. the rdsn framework is used for codes of a service_app,
+    //     not for codes under service_app
+    if (nullptr == _timer) {
+        _timer.reset(new boost::asio::deadline_timer(tools::shared_io_service::instance().ios));
+    }
+    _timer->expires_from_now(boost::posix_time::seconds(renew_gap));
+    _timer->async_wait([this](const boost::system::error_code &err) {
+        if (!err.failed()) {
+            get_credentials();
+            schedule_renew_credentials();
+        } else if (err == boost::system::errc::operation_canceled) {
+            dwarn("the renew credentials timer is cancelled");
+        } else {
+            dassert_f(false, "unhandled error({})", err.message());
+        }
+    });
+}
+
+// switch the code of krb5_xxx function to error_s
```

Comment:
What is the `krb5_xxx function`? maybe " switch krb5_error_code and message to error_s"?